### PR TITLE
Line Item changes

### PR DESCRIPTION
**Pull request by amitshethe**
Changes about line item

### CODE REVIEW

File: app/models/line_item.rb [hunk 1]
```diff
@@ -1,5 +1,10 @@
...
-    belongs_to :cart
-    belongs_to :order,optional: true
-end
+  belongs_to :product
+  belongs_to :cart
+  belongs_to :order,optional: true
+  
+  def total_price
+    self.quantity ||= 0
+    self.quantity * self.product.price
```
Comment (nyadav-deqode) [inline app/models/line_item.rb]: @amitshethe Here self is not required at Line 8.

File: app/controllers/line_items_controller.rb [hunk 2]
```diff
@@ -1,2 +1,52 @@
...
+    if current_cart.products.include?(chosen_product)
+      @line_item = current_cart.line_items.find_by(:product_id => chosen_product)
+      @line_item.quantity += 1
+    else
+      @line_item = LineItem.new
+      @line_item.cart = current_cart
+      @line_item.product = chosen_product
+      @line_item.quantity = 1      
+    end
+    @line_item.save!
```
Comment (nyadav-deqode) [inline app/controllers/line_items_controller.rb:16]: @amitshethe save! raises exception if any error occurred.
Handle if any error occurred it should show the error.

File: app/controllers/line_items_controller.rb [hunk 3]
```diff
@@ -1,2 +1,52 @@
...
+    
+  def destroy
+    @line_item.destroy
+    redirect_to cart_path(@current_cart)
+  end  
+    
+  def add_quantity
+    @line_item.quantity ||= 0
+    @line_item.quantity += 1
+    @line_item.save!
```
Comment (nyadav-deqode) [inline app/controllers/line_items_controller.rb:28]: @amitshethe save! raises exception if any error occurred.
Instead use update_attribute

File: app/controllers/line_items_controller.rb [hunk 4]
```diff
@@ -1,2 +1,52 @@
...
+    params.require(:line_item).permit(:quantity,:product_id, :cart_id)
+  end
+
+  def find_item
+    @line_item = LineItem.find(params[:id])
+  end      
+    
 end
+    
+
```
Comment (nyadav-deqode) [inline app/controllers/line_items_controller.rb:51]: Remove extra lines

File: app/models/line_item.rb [hunk 5]
```diff
@@ -1,5 +1,10 @@
 class LineItem < ApplicationRecord
-    belongs_to :product
-    belongs_to :cart
-    belongs_to :order,optional: true
-end
+  belongs_to :product
+  belongs_to :cart
+  belongs_to :order,optional: true
```
Comment (nyadav-deqode) [inline app/models/line_item.rb]: Add space before optional: true.

Install rubocop to identify these type of errors

File: app/controllers/line_items_controller.rb [hunk 6]
```diff
@@ -1,2 +1,52 @@
...
+    @line_item.quantity += 1
+    @line_item.save!
+    redirect_to cart_path(@current_cart)
+  end
+      
+  def reduce_quantity
+    if @line_item.quantity > 1
+      @line_item.quantity -= 1
+    end
+    @line_item.save
```
Comment (nyadav-deqode) [inline app/controllers/line_items_controller.rb:36]: use update_attribute instead of save

**Review by nyadav-deqode — CHANGES_REQUESTED** [review]

File: app/models/line_item.rb (hunk 1 above)
Comment (amitshethe) [reply app/models/line_item.rb]: @nyadav-deqode I tried without self keyword it doesn't show the current quantity

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -1,2 +1,21 @@
 class ApplicationController < ActionController::Base
+	before_action :authenticate_user!
+	protect_from_forgery with: :exception
+	before_action :current_cart
+  
+	private
+	def current_cart
+		if session[:cart_id]            
+			cart = Cart.find_by(:id => session[:cart_id])
```
Comment (nyadav-deqode) [inline app/controllers/application_controller.rb]: Correct indentions here. Use 2 spaces instead of 4

**Review by nyadav-deqode — CHANGES_REQUESTED** [review]

**Review by nyadav-deqode — APPROVED** [review]

File: app/controllers/line_items_controller.rb (hunk 2 above)
Comment (amitshethe) [reply app/controllers/line_items_controller.rb:16]: Handled the exception if any comes

File: app/controllers/line_items_controller.rb (hunk 3 above)
Comment (amitshethe) [reply app/controllers/line_items_controller.rb:28]: used update_attribute for adding quantity

File: app/controllers/line_items_controller.rb (hunk 4 above)
Comment (amitshethe) [reply app/controllers/line_items_controller.rb:51]: removed extra spaces

File: app/models/line_item.rb (hunk 5 above)
Comment (amitshethe) [reply app/models/line_item.rb]: installed rubocop

File: app/controllers/line_items_controller.rb (hunk 6 above)
Comment (amitshethe) [reply app/controllers/line_items_controller.rb:36]: used update_attribute for reducing quantity

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (amitshethe) [reply app/controllers/application_controller.rb]: done

File: app/controllers/carts_controller.rb [hunk 8]
```diff
@@ -1,3 +1,5 @@
+# frozen_string_literal: true
+
 class CartsController < ApplicationController
   before_action :intialize_Cart_Object, only: %i[show destroy]
```
Comment (nyadav-deqode) [inline app/controllers/carts_controller.rb:4]: @amitshethe update method name in snake_case and where is show action in this controller?

File: app/controllers/line_items_controller.rb [hunk 9]
```diff
@@ -1,2 +1,52 @@
 class LineItemsController < ApplicationController
+  before_action :find_item, only: %i[ destroy add_quantity reduce_quantity]
```
Comment (nyadav-deqode) [inline app/controllers/line_items_controller.rb:2]: Remove extra space before destory @amitshethe

File: app/controllers/line_items_controller.rb (hunk 2 above)
Comment (nyadav-deqode) [reply app/controllers/line_items_controller.rb:16]: I don't see above change

File: app/controllers/line_items_controller.rb (hunk 3 above)
Comment (nyadav-deqode) [reply app/controllers/line_items_controller.rb:28]: I don't see above change @amitshethe

File: app/controllers/line_items_controller.rb (hunk 6 above)
Comment (nyadav-deqode) [reply app/controllers/line_items_controller.rb:36]: I don't see above change

File: app/controllers/line_items_controller.rb (hunk 4 above)
Comment (nyadav-deqode) [reply app/controllers/line_items_controller.rb:51]: Not attended above

**Pull request merged**